File: src/types/subagent.ts
/**
 * Subagent types for CCS stream rendering.
 * Ported from claude-devtools/src/main/types/chunks.ts
 */

import type { ContentBlock } from '@/lib/jsonl-session-parser'

// =============================================================================
// Session Metrics
// =============================================================================

/** Aggregated metrics for a session or subagent */
export interface SessionMetrics {
  /** Duration in milliseconds */
  durationMs: number
  /** Total tokens (input + output) */
  totalTokens: number
  /** Input tokens */
  inputTokens: number
  /** Output tokens */
  outputTokens: number
  /** Cache read tokens */
  cacheReadTokens: number
  /** Cache creation tokens */
  cacheCreationTokens: number
  /** Number of messages */
  messageCount: number
  /** Estimated cost in USD */
  costUsd?: number
}

// =============================================================================
// Team Colors
// =============================================================================

/** Color set for team member visualization */
export interface TeamColorSet {
  /** Border accent color */
  border: string
  /** Badge background (semi-transparent) */
  badge: string
  /** Text color for labels */
  text: string
}

/** Team metadata attached to subagent */
export interface TeamMetadata {
  teamName: string
  memberName: string
  memberColor: string
}

// =============================================================================
// Process (Subagent)
// =============================================================================

/** Parsed message from subagent JSONL (simplified version) */
export interface SubagentMessage {
  uuid: string
  parentUuid: string | null
  type: 'user' | 'assistant' | 'system'
  timestamp: Date
  role?: string
  content: ContentBlock[] | string
  // Tool calls for assistant messages
  toolCalls?: SubagentToolCall[]
  // Tool results for user messages
  toolResults?: SubagentToolResult[]
  // Enriched tool result payload from JSONL entry (if present)
  toolUseResult?: unknown
}

/** Tool call in subagent */
export interface SubagentToolCall {
  id: string
  name: string
  input: Record<string, unknown>
  result?: string
  isError?: boolean
}

/** Tool result in subagent */
export interface SubagentToolResult {
  toolUseId: string
  content: string
  isError: boolean
}

/** Main session impact tokens */
export interface MainSessionImpact {
  /** Task tool_use input tokens */
  callTokens: number
  /** Task tool_result output tokens */
  resultTokens: number
  /** Total tokens affecting main session */
  totalTokens: number
}

/** Resolved subagent information */
export interface Process {
  /** Agent ID extracted from filename */
  id: string
  /** Path to the subagent JSONL file */
  filePath: string
  /** Parsed messages from the subagent session */
  messages: SubagentMessage[]
  /** When the subagent started */
  startTime: Date
  /** When the subagent ended */
  endTime: Date
  /** Duration in milliseconds */
  durationMs: number
  /** Aggregated metrics for the subagent */
  metrics: SessionMetrics
  /** Task description from parent Task call */
  description?: string
  /** Subagent type from Task call (e.g., "Explore", "Plan") */
  subagentType?: string
  /** Whether executed in parallel with other subagents */
  isParallel: boolean
  /** The tool_use ID of the Task call that spawned this */
  parentTaskId?: string
  /** Whether this subagent is still in progress */
  isOngoing?: boolean
  /** Tokens consumed in main session context */
  mainSessionImpact?: MainSessionImpact
  /** Team metadata - present when this subagent is a team member */
  team?: TeamMetadata
  /** Link confidence from resolver matching */
  linkMethod?: 'result' | 'team' | 'fallback'
}

// =============================================================================
// Display Items
// =============================================================================

/** Subagent display item for rendering */
export interface SubagentDisplayItem {
  type: 'subagent'
  subagent: Process
}

/** Teammate message display item */
export interface TeammateMessageDisplayItem {
  type: 'teammate_message'
  teammateMessage: TeammateMessage
}

/** Parsed teammate message content */
export interface TeammateMessage {
  teammateId: string
  color?: string
  summary?: string
  content: string
  timestamp: Date
}

// =============================================================================
// Raw Subagent Data (from Rust)
// =============================================================================

/** Raw subagent data returned from Rust backend */
export interface RawSubagentData {
  id: string
  file_path: string
  raw_messages: string
  start_time: string
  end_time: string
  duration_ms: number
  message_count: number
  first_user_message: string | null
  is_warmup: boolean
  is_compact: boolean
}

File: src-tauri/src/models/subagent.rs
//! Subagent models for IPC between Rust and TypeScript.
//! Matches types in src/types/subagent.ts

use serde::{Deserialize, Serialize};

/// Aggregated metrics for a subagent session
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionMetrics {
    pub duration_ms: u64,
    pub total_tokens: u64,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_read_tokens: u64,
    pub cache_creation_tokens: u64,
    pub message_count: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost_usd: Option<f64>,
}

/// Team metadata for team member subagents
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamInfo {
    pub team_name: String,
    pub member_name: String,
    pub member_color: String,
}

/// Main session impact tokens
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MainSessionImpact {
    pub call_tokens: u64,
    pub result_tokens: u64,
    pub total_tokens: u64,
}

/// Raw subagent data returned from file parsing
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RawSubagentData {
    pub id: String,
    pub file_path: String,
    pub raw_messages: String, // JSONL content as string
    pub start_time: String,   // ISO 8601
    pub end_time: String,     // ISO 8601
    pub duration_ms: u64,
    pub message_count: usize,
    pub first_user_message: Option<String>,
    pub is_warmup: bool,
    pub is_compact: bool,
}
